### PR TITLE
create client for new distributed rmb 

### PR DESCRIPTION
**Pull request by amiraabouhadid**
issue #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
...
+This is a `Typescript` sdk that can be used to build both **services**, and **clients**
+that can talk over the `rmb`.
+
+[RMB](https://github.com/threefoldtech/rmb-rs) is a message bus that enable secure
+and reliable `RPC` calls across the globe.
+
+`RMB` itself does not implement an RPC protocol, but just the secure and reliable messaging
+hence it's up to server and client to implement their own data format.
+
+## Pre Requirements
```
Comment (xmonader) [inline README.md:11]: That's too complicated already, the end user shouldn't be bothered by deploying a relay and running it, please link to one of the deployed relay and move that part of running your own relay to a separate page or flag it as advanced, but definitely not a prerequisite

File: README.md [hunk 2]
```diff
@@ -1 +1,22 @@
...
+You need to run an `rmb-relay` locally. 
+
+### Direct Client
```
Comment (xmonader) [inline README.md:14]: For simplicity let's just call it RMB client instead and not mention the otherflow completely, maybe in a separate page instead

File: README.md (hunk 1 above)
Comment (amiraabouhadid) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (amiraabouhadid) [reply README.md]: done

File: lib/direct.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+}
+export async function createDirectClient(url: string, session: string, mnemonics: string, keyType: string) {
+    // create client
+    const client = await newDirectClient(url, session, mnemonics, keyType);
+
+    return client;
+
+}
+export function sendDirectRequest(client: directClientInterface, socket: ReconnectingWebSocket, requestCommand: string, requestData: any[], destinationTwinId: number) {
```
Comment (AhmedHanafy725) [inline lib/direct.ts:67]: does it a must to have the `requestData` as list? can't send an object (the use case is sending the deployment object to the node)?

File: lib/identity.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import Keyring from '@polkadot/keyring';
+export enum KeypairType {
+    sr25519 = "sr25519",
+    ed25519 = "ed25519"
+}
+export function createIdentity(mnemonics: string, accountType: string) {
```
Comment (AhmedHanafy725) [inline lib/identity.ts:6]:
```suggestion
export function createIdentity(mnemonics: string, accountType: KeypairType) {
```

File: package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    "nodemon": "^2.0.20",
+    "protoc-gen-ts": "^0.8.5",
+    "reconnecting-websocket": "^4.4.0",
+    "socket.io-client": "^4.5.4",
+    "ts-node": "^10.9.1",
+    "ts-protoc-gen": "^0.15.0",
+    "typescript": "^4.9.4",
+    "uuid": "^9.0.0",
+    "websocket-reconnect": "^1.0.7",
+    "ws": "^8.12.0"
```
Comment (AhmedHanafy725) [inline package.json:30]: there are a lot of websockets packages here, do we need all of them?

File: package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  "dependencies": {
+    "@cuillere/core": "^0.13.0",
+    "@polkadot/api": "^9.11.3",
+    "@polkadot/keyring": "^10.2.6",
+    "@polkadot/util-crypto": "^10.2.6",
+    "@protobuf-ts/plugin": "^2.8.2",
+    "base64url": "^3.0.1",
+    "bip39": "^3.0.4",
+    "bytebuffer": "^5.0.1",
+    "express": "^4.18.2",
```
Comment (AhmedHanafy725) [inline package.json:20]: do we need `express` package?

File: package.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    "nodemon": "^2.0.20",
```
Comment (AhmedHanafy725) [inline package.json:21]: I think `nodemon` is a dev dependency

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    "experimentalDecorators": true,
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "forceConsistentCasingInFileNames": true,
+    "useDefineForClassFields": true,
+    "sourceMap": true,
+    "baseUrl": "./lib",
+  },
+  "include": ["lib/*.ts", "lib/types/*.d.ts", "examples/direct/*.ts"],
```
Comment (AhmedHanafy725) [inline tsconfig.json:17]: why do we add the examples in the build?

File: package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "rmb-sdk-ts",
+  "version": "1.0.0",
+  "main": "index.js",
```
Comment (AhmedHanafy725) [inline package.json:4]: is that file `index.js` exist in the client

File: README.md [hunk 10]
```diff
@@ -1,22 +1,48 @@
-# Introduction
-This is a `Typescript` sdk that can be used to build both **services**, and **clients**
+## RMB TS CLIENT
```
Comment (xmonader) [inline README.md:1]: Why screaming?

File: lib/direct.ts (hunk 3 above)
Comment (amiraabouhadid) [reply lib/direct.ts:67]: as protobuf types require string or Uint8Array any type that can be json.stringify then converted to a buffer will do

File: examples/direct/main.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+
+async function main() {
+
+    // create client 
+    const client = await createDirectClient(`wss://relay.dev.grid.tf/`, "test_client", "drama govern gossip audit mixed silent voice mule wonder protect latin idea", 'sr25519');
+    // connect socket
+    const socket = connect(client);
+    // send request
+    const requestID = sendDirectRequest(client, socket, "calculator.add", [10.6, 20], 1292);
```
[the remark this answers is not in the document]
Comment (amiraabouhadid) [reply examples/direct/main.ts]: Can you give more details about what a request to a node should look like? params?

**Review by AhmedHanafy725 — COMMENTED** [review]
please remove the dist files from the pr and add the dist directory to gitignore

File: lib/client.ts [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+class Client {
+    signer!: KeyringPair;
+    source: Address = new Address();
+    twinId: number = 0;
+    url: string = "";
+    responses;
+    con!: ReconnectingWebSocket;
+
+
+    constructor() {
```
Comment (AhmedHanafy725) [inline lib/client.ts:25]: let's set the mnemonics, relayUrl, chainUrl, and keyPairType in the constructor

File: lib/client.ts [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+import ReconnectingWebSocket from "reconnecting-websocket";
+import { Address, Envelope, Request, Response, Error } from "./types/lib/types";
+import { waitReady } from '@polkadot/wasm-crypto';
+import { v4 as uuidv4 } from 'uuid';
+import { ApiPromise, Keyring, WsProvider } from '@polkadot/api'
+import { KeypairType } from "@polkadot/util-crypto/types";
+import crypto from 'crypto';
+import base64url from "base64url";
+import Ws from 'ws';
+const CHAIN_WS = "wss://tfchain.dev.grid.tf/ws";
```
Comment (AhmedHanafy725) [inline lib/client.ts:11]: the urls should be configurable

File: package.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "dependencies": {
+    "@polkadot/api": "^9.11.3",
+    "@polkadot/keyring": "^10.2.6",
+    "@polkadot/util-crypto": "^10.2.6",
+    "base64url": "^3.0.1",
+    "google-protobuf": "^3.21.2",
+    "protoc-gen-ts": "^0.8.5",
+    "reconnecting-websocket": "^4.4.0",
+    "ts-node": "^10.9.1",
```
Comment (AhmedHanafy725) [inline package.json:18]: I think ts-node should be in the dev dependencies

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

File: lib/envelope.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
+import { v4 as uuidv4 } from 'uuid';
+import { Address, Request, Envelope, Error, Response } from './types/lib/types';
+import crypto from 'crypto';
```
Comment (AhmedHanafy725) [inline lib/envelope.ts:3]: crypto is a nodejs package, can we replace it with another package that works on the browser? (maybe crypto-js)

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**